### DISCUSSION TITLE
Airflow scheduler spawn zombie process

### DISCUSSION THREAD

**Discussion by maciejkul** [apache/airflow#25225] [General] [answered]
### Apache Airflow version

2.3.2

### What happened

Airflow scheduler spawn zombie process. I get more then 5000 zombie in daily interval. As workaround, I restart airflow scheduler container daily.

### Below some information from debug
* i don't have any queued DAGs (I don't see any in webserver).
* i cannot find any Zombie PID in airflow logs
* airflow update from 2.2 to 2.3.2 doesn't change anything
* zombie process just "appear". I don't see any relevance.
* i get same issue when all DAGs are disabled
* issue persist only in scheduler
* i don't test daemon mode
* i have 210 dags file

Non zombie process from scheduler container:
```
deploy@50393bd86f1f:/usr/local/airflow$ ps aux | grep -v Z
deploy+     1  6.5  0.1 1081432 194652 ?      Ssl  Jul20  90:49 /usr/local/bin/python3.8 /usr/local/bin/airflow scheduler
deploy+    83  2.7  0.1 937152 158472 ?       S    Jul20  38:12 airflow scheduler -- DagFileProcessorManager
deploy+  1665  1.0  0.0   8060  2288 pts/0    Ss   08:23   0:00 bash
deploy+  1689 13.0  0.1 967880 169384 ?       Sl   08:23   0:00 airflow scheduler - DagFileProcessor /usr/local/airflow/dags/load_009_utlan.py
deploy+  1696  0.0  0.1 939976 162012 ?       S    08:23   0:00 airflow scheduler - DagFileProcessor /usr/local/airflow/dags/confluence_publisher.py
deploy+  1700  0.0  0.0  11560  1616 pts/0    R+   08:23   0:00 ps aux
```

Zombie count:
```
deploy@50393bd86f1f:/usr/local/airflow$ ps aux | grep  Z | wc -l
13552
```
Zombie process (not all):
```
deploy+ 32704  0.0  0.0      0     0 ?        Z    02:25   0:00 [sh] <defunct>
deploy+ 32707  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32713  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32714  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32716  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32718  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32720  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32721  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32724  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32727  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32728  0.0  0.0      0     0 ?        Z    03:20   0:00 [sh] <defunct>
deploy+ 32733  0.0  0.0      0     0 ?        Z    00:31   0:00 [sh] <defunct>
deploy+ 32736  0.0  0.0      0     0 ?        Z    00:31   0:00 [sh] <defunct>
deploy+ 32738  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32740  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32741  0.0  0.0      0     0 ?        Z    03:20   0:00 [sh] <defunct>
deploy+ 32742  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32744  0.0  0.0      0     0 ?        Z    06:02   0:00 [sh] <defunct>
deploy+ 32746  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32748  0.0  0.0      0     0 ?        Z    02:25   0:00 [sh] <defunct>
deploy+ 32750  0.0  0.0      0     0 ?        Z    01:30   0:00 [sh] <defunct>
deploy+ 32751  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32752  0.0  0.0      0     0 ?        Z    06:49   0:00 [sh] <defunct>
deploy+ 32756  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32757  0.0  0.0      0     0 ?        Z    06:02   0:00 [sh] <defunct>
deploy+ 32758  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32760  0.0  0.0      0     0 ?        Z    01:30   0:00 [sh] <defunct>
deploy+ 32761  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32763  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
deploy+ 32764  0.0  0.0      0     0 ?        Z    06:49   0:00 [sh] <defunct>
deploy+ 32767  0.0  0.0      0     0 ?        Z    Jul20   0:00 [sh] <defunct>
```

Zombie parents (not all again):
```
deploy@50393bd86f1f:/usr/local/airflow$ ps -o ppid= -p 32767
    1
deploy@50393bd86f1f:/usr/local/airflow$ ps -o ppid= -p 32757
    1
```

### What you think should happen instead

Wrong subprocess handling in airflow scheduler code.

### Operating System

Debian GNU/Linux 10 (buster) - Docker container

### Versions of Apache Airflow Providers

deploy@50393bd86f1f:/usr/local/airflow/dags$ pip freeze | grep apache-
apache-airflow==2.3.2
apache-airflow-providers-celery==3.0.0
apache-airflow-providers-ftp==3.0.0
apache-airflow-providers-http==3.0.0
apache-airflow-providers-imap==3.0.0
apache-airflow-providers-mongo==3.0.0
apache-airflow-providers-postgres==5.0.0
apache-airflow-providers-sqlite==3.0.0

### Deployment

Docker-Compose

### Deployment details

We use docker-compose with docker swarm. Container is behind proxy.

### Anything else

Issue persist every time

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Can you pleasupgrade to 2.3.3  and check if it's fixed there? There were  a number of fixes implemented there and the easiest way to check is to upgrade (whch you should do anyway as it contains important bugfixes).

Please let us know if it occurs after you do.

**Comment by potiuk** [comment]
Converting to discussion until we see more evidences from latest version

**Comment by maciejkul** [comment]
Hi, i updated airflow to version 2.3.3. Issue still persist.

**Reply by potiuk** [reply]
How do you run scheduiler and why don't your run scheduler in daemon mode? Daemon mode is exactly foreseen for the situation where you need to run airflow in unattended mode by some kind of systemd or similar. Is there a reason why you do not run it in daemon mode and can you see if the same happens if you DO use daemon mode?

**Reply by maciejkul** [reply]
Well, in dockerfile from repository or docker compose from documentation scheduler isn't run in daemon mode. When i run scheduler in container in deamon mode, container exit with completed status, then new container spawned and also quit with completed status (spawn loop)

https://airflow.apache.org/docs/apache-airflow/2.3.3/docker-compose.yaml
https://github.com/apache/airflow/blob/main/Dockerfile

**Reply by potiuk** [reply]
How do you run the airflow scheduler container? 

Are you using the [entrypoint](https://airflow.apache.org/docs/docker-stack/entrypoint.html) that is there, or are you inventing some other way of running airlfow? 

My guess is that likely you made the mistake of running airflow directly which you should never, ever do - you should use dumb init instead (see the docs linked, they explain a lot about it).

Most likely the way you run it is that you run python process directly, making airflow an "init" process. Airflow is not an 'init" process. Proper init process is always needed in order to reap zombies, and my guess is that you changed the entrypoint rather than command to execute and that is causing the problem you observe.

You can also read why it matters to have a proper init process here https://stackoverflow.com/questions/49162358/docker-init-zombies-why-does-it-matter  (for example - there are plenty of sources available).

**Reply by maciejkul** [reply]
Dockerfile and related script below (i had to remove part of information). We also use custom bash endpoint script.
```
FROM myrepo.com/python:0.0.1
ARG AIRFLOW_VERSION=2.3.3
ARG AIRFLOW_USER_HOME=/usr/local/airflow
ARG AIRFLOW_DEPS=""
ARG PYTHON_DEPS=""
ENV AIRFLOW_HOME=${AIRFLOW_USER_HOME}

USER root

RUN useradd -ms /bin/bash -d ${AIRFLOW_USER_HOME} airflow

ADD requirements.txt /requirements.txt
ADD constraints-3.8.txt /constraints-3.8.txt
RUN pip3.8 install -r /requirements.txt
RUN pip3.8 install -c /constraints-3.8.txt  \
    apache-airflow[crypto,celery,postgres]==${AIRFLOW_VERSION} \
    apache-airflow-providers-mongo

COPY entrypoint.sh /entrypoint.sh

RUN ["chmod", "777", "/entrypoint.sh"]

COPY airflow.cfg ${AIRFLOW_USER_HOME}/airflow.cfg

COPY webserver_config.py ${AIRFLOW_USER_HOME}/

EXPOSE 8090 5555 8793

WORKDIR ${AIRFLOW_USER_HOME}
ENTRYPOINT ["/entrypoint.sh"]
```

entrypoint.sh
```
#!/usr/bin/env bash
export \
  AIRFLOW_HOME \
  AIRFLOW__CORE__EXECUTOR \
  AIRFLOW__CORE__FERNET_KEY \
  AIRFLOW__CORE__LOAD_EXAMPLES \
export AIRFLOW__CELERY__BROKER_URL="redis://myredis:5432/10"
exec airflow scheduler
```

**Reply by potiuk** [reply]
Just read the docs. You should not run bash script as entrypoint - you should use dumb-init or similar. This is what creates zombies.

**Reply by potiuk** [reply]
This is Airflow's entrypoint.

```
ENTRYPOINT ["/usr/bin/dumb-init", "--", "/entrypoint"]
```

This is an often mistake people make - using bash script as entrypoint is WRONG. Bash process is not "init' process. It does not propagate signals properly (it swallows them) and does not reap zombies properly. So what you see is entirely expected.

Look it up, there are plenty of articles available on top of those I already pasted here.

**Reply by mavenzer** [reply]
My docker file is almost similar to the above docker file but my  ```AIRFLOW_HOME= /opt/airflow```   when I tried to update the dockerFile with your command  i.e. 
``` ENTRYPOINT ["/usr/bin/dumb-init", "--", "/entrypoint"] ```

I'm getting the following error :
``` container_linux.go:290 :starting container  process caused "exec: \"/usr/bin/dumb-init\" no such file or directory ```

PS :
My airflow instance is also generating lots of zombie processes. I'm also using entrypoint.sh scripts from [here](https://github.com/puckel/docker-airflow/blob/master/script/entrypoint.sh).

**Reply by potiuk** [reply]
Probably you just don't have dumb-init in your image. Make sure you have it.

**Reply by potiuk** [reply]
BTW. using an image that was last updated > 2 years ago and not updated for Airflow 2 is a root of your problem. Get rid of the problem and switch to Airlfow Official image.

**Reply by mavenzer** [reply]
I have installed **dumb-init** as a python library still, it shows no such file or directory when I deploy the webserver. Any possible way to fix it? As in my instance, it also creates lots of zombie processes.

**Reply by mavenzer** [reply]
Do you think entrypoint.sh is not correct for my case? Can you suggest any entrypoint.sh which is compatible with Postgres backend?

**Reply by Taragolis** [reply]
> I have installed dumb-init as a python library still, it shows no such file or directory when I deploy the webserver.

Since you've mentioned that you use Debian, I assume that your image based on debian, so just install it by apt:

```dockerfile
RUN apt-get update \
    && apt-get install -y --no-install-recommends dumb-init
```

> Do you think entrypoint.sh is not correct for my case? Can you suggest any entrypoint.sh which is compatible with Postgres backend?

Just a silly question, why you not use custom image based on Official Airflow Image:
 * https://airflow.apache.org/docs/docker-stack/build.html

It already has [this file as entrypoint](https://github.com/apache/airflow/blob/main/scripts/docker/entrypoint_prod.sh), @potiuk correct me if I wrong.

**Reply by potiuk** [reply]
Absolutely agree with @Taragolis - this is precisely the problems you will have @mavenzer if you opt for your solution. 

To be quite honest and direct - your approach is a bit strange and rather non-empathetic.

First you first shoot yourself in the foot by not using the official/reference image that we have, opting for your own custom version and now you seek for solving your problem among the people who not only recommend you a better solution, but also spent awfully lot of time on testing, and [documenting](https://airflow.apache.org/docs/docker-stack/index.html) the image that has all those problems solved. The documentation specially prepared for people like you so that they do not take individually the time of those people (because with the scale we have we cannot individually help all people who have problems - that's why we prepared the documentation.

And now you ask for even more time of those people to try to resolve your custom solution problems that apparently outgrew your decision  ("I want to manage the image myself") and skills to solve them on your own.

This is not a very good idea and not very empathetic towards those people.

Switch to the reference image as a base.

**Reply by mavenzer** [reply]
@maciejkul   Did your zombie processes issue got resolved by using dumb-init?

**Reply by mavenzer** [reply]
@potiuk Totally agreed with your point its created a hell of lot of issues, But it was a technical debt incurred over time and we intend to fix it very quickly.

**Reply by potiuk** [reply]
@mavenzer then pay someone for fixing it very quickly

**Reply by potiuk** [reply]
People here help in their free time. If you have a problem that you want to solve by tapping into that rather than using resources that are available and recommended, you have to bear the consequences.

**Reply by potiuk** [reply]
You have to pay your debts you know, rather than asking others to pay your debts.

**Reply by mavenzer** [reply]
@potiuk  I appreciate your criticism but we are boot-strapped low on cash startup.  But I really want to support the Apache community once we have sufficient breathing room. Thanks for your efforts again.

**Reply by potiuk** [reply]
Yep. I know it's easier to ask for other volunteers time, rather than volunteer your own time to solve the problem of your own startup. 
Totally and empathically understand that. 

But it does not mean it's moral and good.

**Comment by maciejkul** [comment]
@potiuk @mavenzer 
Hello, sorry for long time response. I was out of computer
Thank you for your time and advices, you were totally right. Using dumb-init resolved my headache.

Have nice time guys!

**Comment by maciejkul** [comment, marked as answer]
Copy @potiuk answer to mark it as resolution

```
This is Airflow's entrypoint.

ENTRYPOINT ["/usr/bin/dumb-init", "--", "/entrypoint"]
This is an often mistake people make - using bash script as entrypoint is WRONG. Bash process is not "init' process. It does not propagate signals properly (it swallows them) and does not reap zombies properly. So what you see is entirely expected.

Look it up, there are plenty of articles available on top of those I already pasted here.
```
